src/Common/Commands.Common.Authentication/Factories/ClientFactory.cs

```diff
@@ -32,17 +32,20 @@ public class ClientFactory : IClientFactory
         private static readonly char[] uriPathSeparator = { '/' };
 
         private Dictionary<Type, IClientAction> _actions;
+        private HashSet<ProductInfoHeaderValue> _headers;
         private OrderedDictionary _handlers;
         private ReaderWriterLockSlim _actionsLock;
         private ReaderWriterLockSlim _handlersLock;
+        private ReaderWriterLockSlim _headersLock;
 
         public ClientFactory()
         {
             _actions = new Dictionary<Type, IClientAction>();
             _actionsLock = new ReaderWriterLockSlim();
-            UserAgents = new HashSet<ProductInfoHeaderValue>();
             _handlers = new OrderedDictionary();
             _handlersLock = new ReaderWriterLockSlim();
+            _headers = new HashSet<ProductInfoHeaderValue>();
+            _headersLock = new ReaderWriterLockSlim();
```

Comment:
Should we be making these instance resource thread safe or rather make static Microsoft.Azure.Commands.Common.Authentication.AzureSession class thread safe? Because I see almost all access to ClientFactory class is via AzureSession.ClientFactory static class.